src/misc/cross.cpp

```diff
@@ -416,6 +416,29 @@ FILE *fopen_wrap(const char *path, const char *mode) {
 	return fopen(path,mode);
 }
 
+// A helper for fopen_wrap that will fallback to read-only if read-write isn't possible.
+// In the fallback case, is_readonly argument is toggled to true.
+// In all cases, a pointer to the file is returned (or nullptr on failure).
+FILE *fopen_wrap_ro_fallback(const std::string filename, bool &is_readonly)
+{
+	// Try with the requested permissions
+	const auto permissions = is_readonly ? "rb" : "rb+";
+	FILE *fp = fopen_wrap(filename.c_str(), permissions);
+
+	// Fallback to read-only?
+	const bool failed_opening_rw = (!fp && !is_readonly);
+	if (failed_opening_rw) {
+		fp = fopen_wrap(filename.c_str(), "rb");
+		if (fp) {
+			is_readonly = true;
+			LOG_INFO("FILESYSTEM: Opened %s read-only per host-filesystem permissions",
+			         filename.c_str());
+		}
+	}
+	// Note: if failed, the caller should provide a context-specific failure message
+	return fp;
```

Comment:
I looked at this a bit more.. maybe this is a bit easier to read? (please modify/improve :+1: )

```c++
	// Try with the requested permissions
	const auto requested_perms = is_readonly ? "rb" : "rb+";
	FILE *fp = fopen_wrap(path, requested_perms);
	if (fp || is_readonly) {
		return fp;
	}

	// Fallback to read-only
	assert(!fp && !is_readonly);
	fp = fopen_wrap(path, "rb");
	if (fp) {
		is_readonly = true;
		LOG_INFO("FILESYSTEM: Opened %s read-only per host filesystem permissions",
		         path);
	}
	// Note: if failed, the caller should provide a context-specific message
	return fp;
```

@shermp - can you see any improvements?

Reply:
No, I can't.

Reply:
There is now two ifs instead of three, which is a great improvement. I don't see further improvement either.